Src/LiquidProjections/Dispatcher.cs

```diff
@@ -18,32 +20,85 @@ public Dispatcher(IEventStore eventStore)
+                transactions, new SubscriptionInfo
+                {
+                    Id = id
+                }), id);
         }
 
+#pragma warning disable 618
         public Dispatcher(CreateSubscription createSubscription)
+#pragma warning restore 618
+        {
+            this.createSubscription = (checkpoint, handler, id) => createSubscription(checkpoint, transactions => handler(
+                transactions, new SubscriptionInfo
+                {
+                    Id = id
+                }), id);
+        }
+        public Dispatcher(CreateSubscriptionWithInfo createSubscription)
         {
             this.createSubscription = createSubscription;
         }
 
-        public IDisposable Subscribe(long? checkpoint, Func<IReadOnlyList<Transaction>, Task> handler, string subscriptionId)
+        public IDisposable Subscribe(long? lastProcessedCheckpoint, Func<IReadOnlyList<Transaction>, Task> handler, string subscriptionId)
+        {
+            return Subscribe(lastProcessedCheckpoint, (transactions, context) => handler(transactions), new SubscriptionOptions
+            {
+                Id = subscriptionId
+            });
+        }
+
+        public IDisposable Subscribe(long? lastProcessedCheckpoint, Func<IReadOnlyList<Transaction>, Task> handler)
+        {
+            return Subscribe(lastProcessedCheckpoint, (transactions, context) => handler(transactions), new SubscriptionOptions());
+        }
+
+        public IDisposable Subscribe(long? lastProcessedCheckpoint, Func<IReadOnlyList<Transaction>, SubscriptionInfo, Task> handler, SubscriptionOptions options = null)
         {
+            if (options == null)
+            {
+                options = new SubscriptionOptions();
+            }
+
             if (handler == null)
             {
                 throw new ArgumentNullException(nameof(handler));
             }
 
             var subscriptionMonitor = new object();
-            IDisposable subscription = null;
-
             lock (subscriptionMonitor)
             {
-                subscription = createSubscription(checkpoint,
-                    async transactions =>
+                // NOTE: Try to get the transaction that the subscriber already processed. If it's not there, then we know
+                // the underlying event store is behind and the subscriber can take actions.
+                long? alternateCheckpointToDetectAheadSubscriber = (lastProcessedCheckpoint == null || lastProcessedCheckpoint == 0) ? 
+                    lastProcessedCheckpoint : lastProcessedCheckpoint - 1;
+
+                return createSubscription(alternateCheckpointToDetectAheadSubscriber,
+                    async (transactions, info) =>
                     {
                         try
                         {
-                            await handler(transactions);
+                            if (options.RestartWhenAhead && !transactions.Any())

```

Comment:
We have to modify all the event stores anyway. Otherwise they will not call this delegate with empty batch. So they better call another delegate in that case.